### PR TITLE
Prevent infinite recursion of modules

### LINKED ISSUE #50196: rustdoc stack overflow on stable (1.25.0) and nightly (ac3c2288f)
I got a stack overflow during `cargo doc` on [my crate](https://github.com/RazrFalcon/resvg/tree/909b16121a91d4e647bc310aa25ef74eac897f0c):

```
% cargo doc
 Documenting resvg v0.2.0 (file:///media/data/Programming/Projects/rust/resvg)

thread 'rustc' has overflowed its stack
fatal runtime error: stack overflow
error: Could not document `resvg`.

Caused by:
  process didn't exit successfully: `rustdoc --crate-name resvg src/lib.rs -o /media/data/Programming/Projects/rust/resvg/target/doc -L dependency=/media/data/Programming/Projects/rust/resvg/target/debug/deps --extern failure=/media/data/Programming/Projects/rust/resvg/target/debug/deps/libfailure-0a3bba1278ba0cc9.rmeta --extern log=/media/data/Programming/Projects/rust/resvg/target/debug/deps/liblog-536ed3d82960b225.rmeta --extern euclid=/media/data/Programming/Projects/rust/resvg/target/debug/deps/libeuclid-64eb361a44c7fcf2.rmeta --extern usvg=/media/data/Programming/Projects/rust/resvg/target/debug/deps/libusvg-c79e4aac850c4257.rmeta --extern lyon_geom=/media/data/Programming/Projects/rust/resvg/target/debug/deps/liblyon_geom-0b70e94ea051fc6c.rmeta` (signal: 6, SIGABRT: process abort signal)
```

```
% rustc -V
rustc 1.25.0 (84203cac6 2018-03-25)
% rustc +nightly -V
rustc 1.27.0-nightly (ac3c2288f 2018-04-18)
% uname -a
Linux razr 4.9.76-gentoo-r1 #1 SMP PREEMPT Fri Mar 23 15:37:06 EET 2018 x86_64 Intel(R) Core(TM) i5-3570K CPU @ 3.40GHz GenuineIntel GNU/Linux
```

### PR DESCRIPTION
**Pull request by GuillaumeGomez**
Fixes #50196.

r? @QuietMisdreavus

### CODE REVIEW

File: src/librustdoc/clean/inline.rs [hunk 1]
```diff
@@ -401,7 +401,7 @@ fn build_module(cx: &DocContext, did: DefId) -> clean::Module {
         for &item in cx.tcx.item_children(did).iter() {
             let def_id = item.def.def_id();
             if item.vis == ty::Visibility::Public {
-                if !visited.insert(def_id) { continue }
+                if did == def_id || !visited.insert(def_id) { continue }
```
[the remark this answers is not in the document]
Comment (GuillaumeGomez) [reply src/librustdoc/clean/inline.rs]: Good point. I'll store every def id instead to be sure the item isn't one of its parents.

**Comment by GuillaumeGomez** [comment]
Updated.

**Comment by QuietMisdreavus** [comment]
Can you change your test (or add a new one) to use the sample @ollie27 provided?

**Comment by shepmaster** [comment]
Ping from triage, @GuillaumeGomez ! You have some feedback to address!

**Comment by GuillaumeGomez** [comment]
Indeed, will update *soon*. :)

**Comment by GuillaumeGomez** [comment]
Updated.

**Comment by QuietMisdreavus** [comment]
Thanks!

@bors r+

**Pull request merged**